tripy/examples/segment-anything-model-v2/sam2/modeling/memory_encoder.py

```diff
@@ -0,0 +1,198 @@
+    """
+
+    def __init__(
+        self,
+        dim,
+        kernel_size=7,
+        padding=3,
+        drop_path=0.0,
+        layer_scale_init_value=1e-6,
+        use_dwconv=True,
+        dtype="float32",
+    ):
+        super().__init__()
+        self.dtype = getattr(tp, dtype)
+        self.dwconv = tp.Conv(
+            dim,
+            dim,
+            kernel_dims=(kernel_size, kernel_size),
+            padding=((padding, padding), (padding, padding)),
+            groups=dim if use_dwconv else 1,
+            dtype=self.dtype,
+        )  # depthwise conv
+        self.norm = LayerNorm2d(dim, eps=1e-6)
+        self.pwconv1 = tp.Linear(dim, 4 * dim, dtype=self.dtype)  # pointwise/1x1 convs, implemented with linear layers
+        self.act = tp.gelu
+        self.pwconv2 = tp.Linear(4 * dim, dim, dtype=self.dtype)
+        self.gamma = tp.ones((dim,), dtype=self.dtype) * layer_scale_init_value
+
+        self.drop_path = Dummy()
+
+    def __call__(self, x):
+        return self.forward(x)
+
+    def forward(self, x):
+        input = x
+        x = self.dwconv(x)
+        x = self.norm(x)
+        x = tp.permute(x, (0, 2, 3, 1))  # (N, C, H, W) -> (N, H, W, C)
+        x = self.pwconv1(x)
+        x = self.act(x)
+        x = self.pwconv2(x)
+        if self.gamma is not None:
+            x = self.gamma * x
+        x = tp.permute(x, (0, 3, 1, 2))  # (N, H, W, C) -> (N, C, H, W)
+
+        x = input + self.drop_path(x)
+        return x
+
+
+class Fuser(tp.Module):
+    def __init__(self, layer, num_layers, dim=None, input_projection=False, dtype="float32"):
+        super().__init__()
+        self.dtype = getattr(tp, dtype)
+        self.proj = Dummy()
+        self.layers = [layer for i in range(num_layers)]
+
+        if input_projection:
+            self.proj = tp.Conv(dim, dim, kernel_dims=(1, 1), dtype=self.dtype)
+
+    def __call__(self, x):
+        return self.forward(x)
+
+    def forward(self, x):
+        x = self.proj(x)
+        for layer in self.layers:
+            x = layer(x)
+        return x
+
+
+class MemoryEncoder(tp.Module):
+    def __init__(
+        self, out_dim, mask_downsampler, fuser, position_encoding, in_dim=256, dtype="float32"  # in_dim of pix_feats
+    ):
+        super().__init__()
+        self.dtype = getattr(tp, dtype)
+
+        self.mask_downsampler = mask_downsampler
+        self.pix_feat_proj = tp.Conv(in_dim, in_dim, kernel_dims=(1, 1), dtype=self.dtype)
+        self.fuser = fuser
+        self.position_encoding = position_encoding
+        self.out_proj = Dummy()
+        if out_dim != in_dim:
+            self.out_proj = tp.Conv(in_dim, out_dim, kernel_dims=(1, 1), dtype=self.dtype)
+
+    def __call__(
+        self,
+        pix_feat: tp.Tensor,
+        masks: tp.Tensor,
+        skip_mask_sigmoid: bool = False,
+    ):
+        return self.forward(pix_feat, masks, skip_mask_sigmoid)
```

Comment:
Why the indirection here?

Reply:
Good catch, it was there in older code to maintain torch compatibility. I can remove it in standalone demo.